### PR TITLE
gh-1762 - check added to check group of element being added is in schema

### PR DESCRIPTION
**Pull request by m55624**

### CODE REVIEW

**Comment by gaffer01** [comment]
Don't think this is the right way to do it.

- This consumes the entire iterable of elements to be added to check whether the groups are in the schema - it then needs to be consumed again to add the elements - this is very expensive and there's no guarantee that the iterable can be consumed twice.
- It iterates through all the elements and if any one has a group in the schema then it says it's fine and will then attempt to add all the elements - you'll still get a NPE if one of the elements has a group that's in the schema and one doesn't.

The check that the group is valid needs to be done per element just before the element is added - if the group is not in the schema then a warning should be logged.

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/ParquetStoreTest.java [hunk 1]
```diff
@@ -16,18 +16,67 @@
...
+                    .aggregate(false)
+                    .build())
+            .entity(TestGroups.ENTITY, new SchemaEntityDefinition.Builder()
+                    .property(TestPropertyNames.PROP_1, TestTypes.PROP_INTEGER)
+                    .vertex(TestTypes.ID_STRING)
+                    .aggregate(false)
+                    .build())
+            .build();
+    private final Edge testEdge = new Edge.Builder()
+            .group(TestGroups.EDGE_2)
```
Comment (p013570) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/ParquetStoreTest.java]: Perhaps rename this to unknownEdge. Similarly for the entity.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/handler/utilities/WriteUnsortedData.java [hunk 2]
```diff
@@ -84,17 +84,17 @@ private void _writeElements(final Iterator<? extends Element> elements) throws O
...
             } else {
                 splitToWriter = new HashMap<>();
                 groupSplitToWriter.put(group, splitToWriter);
             }
             if (schemaUtils.getEntityGroups().contains(group)) {
                 writer = getWriter(splitToWriter, groupToSplitPoints.get(group), ((Entity) element).getVertex(), group, ParquetStoreConstants.VERTEX);
-            } else {
+            } else if (schemaUtils.getEdgeGroups().contains(group)) {
                 writer = getWriter(splitToWriter, groupToSplitPoints.get(group), ((Edge) element).getSource(), group, ParquetStoreConstants.SOURCE);
             }
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/handler/utilities/WriteUnsortedData.java]: Can we have a final else to log the fact that the element is ignored if it's not in schema - probably DEBUG level logging?

File: store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/ParquetStoreTest.java [hunk 3]
```diff
@@ -39,4 +88,36 @@ public void testTraits() throws StoreException {
...
+    }
+
+    @Test
+    public void shouldNotThrowExceptionWhenAddingAMultipleElementsWithGroupsNotInSchema() throws Exception {
+        getGraph().execute(new AddElements.Builder()
+                        .input(testEntity, testEdge)
+                        .build(),
+                new User());
+    }
+
```
Comment (gaffer01) [inline store-implementation/parquet-store/src/test/java/uk/gov/gchq/gaffer/parquetstore/ParquetStoreTest.java]: Can we have one more test to cover the case when one element is not in the schema, and one is in the schema - in this case the one that is in the schema should still be added.

File: store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/handler/utilities/WriteUnsortedData.java (hunk 2 above)
Comment (m55624) [reply store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/handler/utilities/WriteUnsortedData.java]: This is on the four lines below, but at a WARN level.

Comment (gaffer01) [reply store-implementation/parquet-store/src/main/java/uk/gov/gchq/gaffer/parquetstore/operation/handler/utilities/WriteUnsortedData.java]: So it is, ignore my comment.

**Pull request merged**